### PR TITLE
dialects: Add trivial GPU operations.

### PR DESCRIPTION
**Pull request by PapyChacal**
That is:
- BarrierOp
- BlockDimOp
- BlockIdOp
- GlobalIdOp
- GridDimOp
- LaneIdOp
- NumSubgroupsOp
- SetDefaultDeviceOp
- SubgroupIdOp
- SubgroupSizeOp
- ThreadIdOp

### CODE REVIEW

File: xdsl/dialects/gpu.py [hunk 1]
```diff
@@ -127,4 +194,64 @@ def get() -> ModuleEndOp:
...
+
+
+#_GPUAttr has to be registered instead of DimensionAttr and AllReduceOperationAttr here.
+# This is a hack to fit MLIR's syntax in xDSL's way of parsing attributes, without making GPU builtin.
+# Hopefully MLIR will parse it in a more xDSL-friendly way soon, so all that can be factored in proper xDSL
+# atrributes.
+GPU = Dialect([
+    BarrierOp, BlockDimOp, BlockIdOp, GlobalIdOp, GridDimOp, LaneIdOp,
+    ModuleOp, ModuleEndOp, NumSubgroupsOp, SetDefaultDeviceOp, SubgroupIdOp,
+    SubgroupSizeOp, ThreadIdOp
```
Comment (AntonLydike) [inline xdsl/dialects/gpu.py:256]: Nit: Add commas at the end to make yapf have nice newlines

**Review by AntonLydike — APPROVED** [review]
LGTM except for some nitpicking, but that's honestly fine

File: tests/dialects/test_gpu.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
-from xdsl.dialects.builtin import SymbolRefAttr
-from xdsl.dialects.gpu import AllReduceOperationAttr, ModuleEndOp, ModuleOp, DimensionAttr
+from xdsl.dialects.arith import Constant
+from xdsl.dialects.builtin import SymbolRefAttr, i32
+from xdsl.dialects.gpu import AllReduceOperationAttr, BarrierOp, BlockDimOp, BlockIdOp, GlobalIdOp, GridDimOp, LaneIdOp, ModuleEndOp, ModuleOp, DimensionAttr, NumSubgroupsOp, SetDefaultDeviceOp, SubgroupIdOp, SubgroupSizeOp, ThreadIdOp
```
Comment (superlopuh) [inline tests/dialects/test_gpu.py:3]:
```suggestion
import xdsl.dialects.gpu as gpu
```

**Review by superlopuh — APPROVED** [review]

File: tests/dialects/test_gpu.py (hunk 2 above)
Comment (PapyChacal) [reply tests/dialects/test_gpu.py:3]: I like the idea of prefixing the dialects in most places, even everywhere out of the dialect, but I wouldn't naturally do it on the said dialect's unit tests. I kind of like seeing them as in the same namespace (e.g. I would put them in the same namespace in  C++). Does it make sense to you?

File: xdsl/dialects/gpu.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/gpu.py:256]: If that is something we want to enforce, can we make yapf do it?
I hate having to manually check my formatter's formatting TBH.

File: tests/dialects/test_gpu.py (hunk 2 above)
Comment (superlopuh) [reply tests/dialects/test_gpu.py:3]: Sure, although there's a case to be made for it being a client and so shouldn't share the namespace. I'm not really familiar with industry practices in C++, so not sure how it works there.

**Pull request merged**